### PR TITLE
Age model max preprocessing

### LINKED ISSUE #2: Convert labels in images into a CSV file
Currently, all the labels for the age model are located in the image file paths themselves in the format

`[age]_[gender]_[race]_[date&time].jpg`.

Instead, to better analyze our dataset, we should extra these labels to create a CSV file with the same headings (age, gender, race, date, and image path).

### LINKED ISSUE #3: Create test set
Create a test set (~20% of images)

### LINKED ISSUE #4: Prepare the data - crop to faces
For the training data, we need to crop the image to just the face. This will occur in pre-processing.

### PR DESCRIPTION
**Pull request by maxklema**
This Pull Request prepares (preprocesses) images from the UTKFace dataset (https://susanqq.github.io/UTKFace/) before they are trained in our CNN model.

### Get the Data

Here, images are pulled from three different tar files and decompressed before being combined into one 'image' folder under `datasets/faces/images`. A total of 24106 (raw images) are scraped.

### Cropping Images & Adding Padding

Many of the images contain backgrounds and other objects outside of the person's face. To get only the face, I use the `face_recognition` library that uses a hog model to crop the raw images to the subject's face. The result is a `n x n` cropped image. However, to keep `n` consistent for all cropped images, which may be of different sizes, I use `Image` and `ImageOps` libraries from `PIL`, specifically the `ImageOps.pad` function which takes an image file, a dimension tuple (224 x 224), a upscaling/downsampling algorithm (Lanzcos), and a padding color (black). 

Each of these images is saved to `datasets/faces/cropped.` To speed up this process, I use the `concurrent.futures` library with `ProcessPoolExecutor` and `as_completed` in conjunction to utilize my multi-core CPU to enable parallelism. Thus, each core acts as a worker and processes one image at a time.

Also - Cropped Image Zip can be found here: https://drive.google.com/file/d/12g152BZ6lwNQ0H5hDkDVG9toGHTEYQN4/view?usp=drive_link

### CNN Age Model

Using an efficient net b0 backbone, I use already pre-trained imagenet weights that are trained to detect different facial features, edges, textures, skin patterns, etc instead of having to train all of this ourselves. I then use batch normalization, dropouts, and a few dense layers to compress the learned age features into a single neuron, which returns the predicted age. All activation functions use ReLU for regression.

### Training

There are two phases of training. The first phase freezes backbone parameters and trains the weights on the dense layers. The second phrase unfreezes the layer layers of the backbone to better adapt to our image data.

### Validation/Testing

The data is divided into roughly ~80% training, 10% validation, and 10% testing. At the end of training, a validation vs training MAE (mean average error) vs. epoch and loss vs. epoch graph is made to show where overfitting/underfitting may occur.

### Testing Model

The model is then tested on the testing data, as well as a few select images to see how it performs.

### CODE REVIEW

**Comment by zeshengchen** [comment]
When the pull request is ready for peer review, please invite ALL other group member to review it.

**Review by Dillstan — APPROVED** [review]

**Review by mylesEdwards — COMMENTED** [review]
Looks Excellent Max!

**Review by IvieXue — APPROVED** [review]
Looks good.

**Review by mylesEdwards — APPROVED** [review]
Ready to merge!
